streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -0,0 +1,231 @@
+                throw new ProcessorStateException(message, e);
+            }
+        }
+
+        @Override
+        protected <R> QueryResult<R> runRangeQuery(final Query<R> query,
+                                                   final PositionBound positionBound,
+                                                   final QueryConfig config) {
+            // throw exception for now to reserve the ability to implement this in the future
+            // without clashing with users' custom implementations in the meantime
+            throw new UnsupportedOperationException("Versioned stores do not support RangeQuery queries at this time.");
+        }
+
+        @Override
+        protected <R> QueryResult<R> runKeyQuery(final Query<R> query,
+                                                 final PositionBound positionBound,
+                                                 final QueryConfig config) {
+            // throw exception for now to reserve the ability to implement this in the future
+            // without clashing with users' custom implementations in the meantime
+            throw new UnsupportedOperationException("Versioned stores do not support KeyQuery queries at this time.");
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        protected Serde<ValueAndTimestamp<V>> prepareValueSerdeForStore(
+            final Serde<ValueAndTimestamp<V>> valueSerde,
+            final SerdeGetter getter
+        ) {
+            if (valueSerde == null) {
+                return new NullableValueAndTimestampSerde<>((Serde<V>) getter.valueSerde());
+            } else {
+                return super.prepareValueSerdeForStore(valueSerde, getter);
+            }
+        }
+    }
+
+    @Override
+    public void put(final K key, final V value, final long timestamp) {
+        internal.put(key, ValueAndTimestamp.makeAllowNullable(value, timestamp));
+    }
+
+    @Override
+    public VersionedRecord<V> delete(final K key, final long timestamp) {
+        final ValueAndTimestamp<V> valueAndTimestamp = internal.delete(key, timestamp);
+        return valueAndTimestamp == null
+            ? null
+            : new VersionedRecord<>(valueAndTimestamp.value(), valueAndTimestamp.timestamp());
+    }
+
+    @Override
+    public VersionedRecord<V> get(final K key) {
+        final ValueAndTimestamp<V> valueAndTimestamp = internal.get(key);
+        return valueAndTimestamp == null
+            ? null
+            : new VersionedRecord<>(valueAndTimestamp.value(), valueAndTimestamp.timestamp());
+    }
+
+    @Override
+    public VersionedRecord<V> get(final K key, final long asOfTimestamp) {
+        final ValueAndTimestamp<V> valueAndTimestamp = internal.get(key, asOfTimestamp);
+        return valueAndTimestamp == null
+            ? null
+            : new VersionedRecord<>(valueAndTimestamp.value(), valueAndTimestamp.timestamp());
+    }
+
+    @Override
+    public String name() {
```

Comment:
It seems we overwrite some method to delegate to `internal.x()` instead of `inner.x()` but I am not sure what the pattern is. Can you elaborate?

Reply:
All methods from `MeteredVersionedKeyValueStore` delegate to `internal`; the methods in `MeteredVersionedKeyValueStore` are purely a wrapper for the internal `MeteredVersionedKeyValueStoreInternal` instance. All the actual logic happens in `MeteredVersionedKeyValueStoreInternal`, which delegates its methods to `inner`, which is the actual inner store (either a changelogging layer or the inner versioned store implementation itself). The only reason that `MeteredVersionedKeyValueStore` implements `WrappedStateStore` at all is because there are places in the DSL code which expect the outermost store layer to be a `WrappedStateStore` in order to call methods such as `setFlushListener()` and `flushCache()`.